python/cuml/linear_model/linear_regression_mg.pyx

```diff
@@ -23,610 +22,474 @@ import ctypes
+            self._freeDoubleD(labels, arr_interfaces_y)
 
-        n_rows = alloc_info[0][0]["shape"][0]
-
-        cdef n_inputs = len(alloc_info)
-
-        cdef int* input_cols
-        cdef int* label_rows
-        cdef int* coef_cols
-        input_cols = <int*>malloc(len(alloc_info*sizeof(int)))
-        label_rows = <int*>malloc(len(alloc_info*sizeof(int)))
-        coef_cols = <int*>malloc(len(alloc_info*sizeof(int)))
-
-        cdef uintptr_t input_ptr
-        n_allocs = len(alloc_info)
-
-        cdef float** input32
-        cdef float** labels32
-        cdef float** coef32
-        cdef double** input64
-        cdef double** labels64
-        cdef double** coef64
-
-        cdef float intercept_f32
-        cdef double  intercept_f64
-        intercept_f32 = 0
-        intercept_f64 = 0
-
-        if dtype == "single":
-            input32 = <float**>malloc(len(alloc_info)*sizeof(float*))
-            labels32 = <float**>malloc(len(alloc_info)*sizeof(float*))
-            coef32 = <float**>malloc(len(alloc_info)*sizeof(float*))
-
-            idx = 0
-            for ary in alloc_info:
-                input_ptr = ary[0]['data'][0]
-                input32[idx] = <float*>input_ptr
-                input_cols[idx] = ary[0]["shape"][1]
-                n_cols = n_cols + ary[0]["shape"][1]
-
-                input_ptr = ary[1]['data'][0]
-                labels32[idx] = <float*>input_ptr
-                label_rows[idx] = ary[1]["shape"][0]
-
-                input_ptr = ary[2]['data'][0]
-                coef32[idx] = <float*>input_ptr
-                coef_cols[idx] = ary[2]["shape"][0]
-
-                idx = idx + 1
-
-            olsFitSPMG(<float**> input32,
-                       <int*> input_cols,
-                       <int> n_rows,
-                       <int> n_cols,
-                       <float**> labels32,
-                       <int*> label_rows,
-                       <float**> coef32,
-                       <int*> coef_cols,
-                       <float*> &intercept_f32,
-                       <bool> self.fit_intercept,
-                       <bool> self.normalize,
-                       <int> n_allocs)
-
-            return intercept_f32
+    def predict(self, X, M, N, partsToRanks, rnk):
+        """
+        Transform function for Linear Regression MG.
+        This not meant to be used as
+        part of the public API.
+        :param X: array of local dataframes / array partitions
+        :param M: total number of rows
+        :param N: total number of cols
+        :param partsToRanks: array of tuples in the format: [(rank,size)]
+        :return: self
+        """
 
-        else:
+        if N != self.n_cols:
+            raise Exception("Number of columns of the X has to match with "
+                            "number of columns of the data was fit to model.")
+
+        arr_interfaces = []
+        for arr in X:
+            X_m, input_ptr, n_rows, n_cols, self.dtype = \
```

Comment:
As mentioned in Ridge, I really think doing this with map_partitions (task-based, embarassingly parallel) rather than by bringing together the data pointers would lead to cleaner code. I think I'm ok deferring that for the next rev though.

Reply:
This is a util that @dantegd is working. @dantegd please correct me if I am wrong. Maybe, that util can do the calculation embarassingly parallel if we can.